### PR TITLE
Rate side configurable

### LINKED ISSUE #1059: Bot uses maker fees on marketable limit orders.
In trading a market maker is a limit order that is far outside of spot price.
Typically an exchange will have positive fee(they pay you) or reduced fee for adding liquidity to the market.

As our limit orders are on spot or close to market price they will incur taker fees not maker. 
The code should reflect this

Limit orders placed at spot price are not treated as maker, or limit order.
They're a 'marketable limit' order, which is different to a limit order, and incurs taker fees as any market order would.

To use limit, maker, fees the limit order should be placed early, left hanging,  far outside of spot.

  ```
 Fee is applied twice because we make a LIMIT_BUY and LIMIT_SELL
        fee = self.exchange.get_fee(symbol=pair, taker_or_maker='maker')
        trade = Trade(
            pair=pair,
            stake_amount=stake_amount,
            amount=amount,
            fee_open=fee,
            fee_close=fee,
            open_rate=buy_limit,
            open_rate_requested=buy_limit,
            open_date=datetime.utcnow(),
            exchange=self.exchange.id,
            open_order_id=order_id
        )
 
```

### LINKED ISSUE #1667: orderbook sell - odd behavior
##  Describe your environment

  * Python Version: 3.7.0
  * CCXT version: 1.18.367 (`pip freeze | grep ccxt`)
  * Branch: Develop
  * Last Commit ID: d66e6510e30c65cd8159104b688ae433b98927c5 
  
## Step 3: Describe the problem:

Using orderbook with the following ask strategy, the bot buys and sells almost immediately with ROI as sell reason.

Configuration:
```
    minimal_roi = {
        "0": 0.01
    }
```

``` json
"ask_strategy":{
    "use_order_book": false,
    "order_book_min": 1,
    "order_book_max": 10
}
```

The relevant code seems to be the following:
https://github.com/freqtrade/freqtrade/blob/71c530590eb4080d901c11075ff577e9dc6296c9/freqtrade/freqtradebot.py#L600-L614 

It walks the ask-side and once it finds a spot matching ROI / sell-signal places the sell there.

In production mode, the order will either be fullfilled once it reaches this limit - or time out (more likely - probably).

In dry/validation mode however it'll produce very odd and misleading results since it assumes that rate reached - if i set my ROI to 10% - and walk 100 steps up the orderbook i may find that filled immediately as well - depending on the pair.

2019-03-16-131755_537x228_scrot

Notice the difference between limit and "current rate" (profit uses limit for the calculation ...).

### Relevant code exceptions or logs:

  ```
2019-03-16 13:14:50,378 - freqtrade.freqtradebot - INFO - Buy signal found: about create a new trade with stake_amount: 0.1 ...
2019-03-16 13:14:50,378 - freqtrade.freqtradebot - INFO - Using Last Ask / Last Price
2019-03-16 13:14:50,692 - freqtrade.rpc.rpc_manager - INFO - Sending rpc message: {'type': buy, 'exchange': 'Binance', 'pair': 'AST/ETH', 'limit': 0.0002841, 'stake_amount': 0.1, 'stake_currency': 'ETH', 'fiat_currency': 'USD'}
2019-03-16 13:14:50,816 - freqtrade.wallets - INFO - Wallets synced ...
2019-03-16 13:14:50,829 - freqtrade.wallets - INFO - Wallets synced ...
2019-03-16 13:14:52,678 - freqtrade.freqtradebot - INFO - Found open order for Trade(id=1, pair=AST/ETH, amount=351.98873636, open_rate=0.00028410, open_since=just now)
2019-03-16 13:14:52,679 - freqtrade.freqtradebot - INFO - Applying fee on amount for Trade(id=1, pair=AST/ETH, amount=351.98873636, open_rate=0.00028410, open_since=just now) failed: myTrade-Dict empty found
2019-03-16 13:14:52,680 - freqtrade.persistence - INFO - Updating trade (id=1) ...
2019-03-16 13:14:52,680 - freqtrade.persistence - INFO - LIMIT_BUY has been fulfilled for Trade(id=1, pair=AST/ETH, amount=351.98873636, open_rate=0.00028410, open_since=just now).
2019-03-16 13:14:52,712 - freqtrade.freqtradebot - INFO - Using order book for selling...
2019-03-16 13:14:53,088 - freqtrade.freqtradebot - INFO -   order book asks top 1: 0.00028410
2019-03-16 13:14:53,089 - freqtrade.freqtradebot - INFO -   order book asks top 2: 0.00028440
2019-03-16 13:14:53,089 - freqtrade.freqtradebot - INFO -   order book asks top 3: 0.00028540
2019-03-16 13:14:53,090 - freqtrade.freqtradebot - INFO -   order book asks top 4: 0.00028570
2019-03-16 13:14:53,090 - freqtrade.freqtradebot - INFO -   order book asks top 5: 0.00028630
2019-03-16 13:14:53,091 - freqtrade.freqtradebot - INFO -   order book asks top 6: 0.00028770
2019-03-16 13:14:53,098 - freqtrade.freqtradebot - INFO - Using order book for selling...
2019-03-16 13:14:53,526 - freqtrade.rpc.rpc_manager - INFO - Sending rpc message: {'type': sell, 'exchange': 'Binance', 'pair': 'AST/ETH', 'gain': 'profit', 'limit': 0.0002877, 'amount': 351.98873636043646, 'open_rate': 0.0002841, 'current_rate': 0.0002841, 'profit_amount': 0.00106589, 'profit_percent': 0.01064827, 'sell_reason': 'roi', 'stake_currency': 'ETH', 'fiat_currency': 'USD'}
2019-03-16 13:14:53,617 - freqtrade.freqtradebot - INFO - executed sell, reason: SellType.ROI
2019-03-16 13:14:53,618 - freqtrade.wallets - INFO - Wallets synced ...
2019-03-16 13:14:53,624 - freqtrade.wallets - INFO - Wallets synced ...

```

### LINKED ISSUE #2637: Wrong usage of bids and asks from orderbook during sell
If `use_order_book` is set to True in the `ask_strategy`, the `handle_trade()` implements its own way to calculate sell_rate -- it uses asks from the orderbook:

https://github.com/freqtrade/freqtrade/blob/aa335d8485cb7099fd5a084f685babd3f4b4d1dd/freqtrade/freqtradebot.py#L606-L626

else it calls `get_sell_rate()`, which also has a branch for calculating sell_rate when `use_order_book` is set to True -- using bid prices from either orderbook or from the ticker. However, this is not used, since `handle_trade()` calculates sell_rate by itself using ask prices. This price (calculated using asks) is then passed to `_check_and_execute_sell()`. The latter passes this sell_rate to `should_sell` and `execute_sell`. The latter finally calls `_notify_sell`. The latter calls `get_sell_rate()` again, which in this case has a chance to calculate the sell rate using bids, so the bid price is reported. In the same time, all logic in `should_sell` and `execute_sell` used the price calculated by `handle_trade` using ask prices.

In my opinion, the logic for defining sell_rate using asks should be imployed (moved, merged somehow) into the generic `get_sell_rate`. Even better, both bid price and ask prices should be fetched from either the ticker or the orderbook and passed to `should_sell` and `execute_sell` -- because it depends on the type of sell order, will it be a sell market (when bids should be used) order or a sell limit order (when asks should be used). Corresponding price (bid or ask, depending on a type of the sell order, market or limit, whether it's sell signal, roi or stoploss, the types may differ) should be used in all calculations inside `should_sell()`, in `stoploss_reached()` and `min_roi_reached()`, same should be reported in `_notify_sell()`

The explicit bug consists now in that `should_sell` logic uses ask prices and `_notify_sell` then reports bid price from the orderbook.

### LINKED ISSUE #2907: get_sell_rate() using bids instead of asks?
Hello,

So, I just noticed that get_sell_rate() method uses bids instead of asks, and I wasn't sure why. Logging this to discuss this as I'm not sure why it's implemented this way. Talking to @hroff-1902 he mentioned that he'd eventually like to see both get_buy_rate and get_sell_rate refactored to be used differently with limit/market orders, so perhaps this is an opportunity to discuss this.

There were also several discussions somewhat related to this made in #1667 and #2637, but I wasn't sure they were entirely focused on using bids in get_sell_rate.

```
def get_sell_rate(self, pair: str, refresh: bool) -> float:
        """
        Get sell rate - either using get-ticker bid or first bid based on orderbook
        The orderbook portion is only used for rpc messaging, which would otherwise fail
        for BitMex (has no bid/ask in fetch_ticker)
        or remain static in any other case since it's not updating.
        :return: Bid rate
        """
        config_ask_strategy = self.config.get('ask_strategy', {})
        if config_ask_strategy.get('use_order_book', False):
            logger.debug('Using order book to get sell rate')
            order_book = self.exchange.get_order_book(pair, 1)
            rate = order_book['bids'][0][0]
        else:
            rate = self.exchange.fetch_ticker(pair, refresh)['bid']
        return rate
```

### LINKED ISSUE #3005: FTX - orderbook no trades
It seems dogehedge pair does not have any orderbook trades in the last 24 hrs, ccxt queries and gets a NoneType which then yields an error. 

Using market order, latest dev

```
2020-03-01 01:17:28,780 - freqtrade.wallets - INFO - Wallets synced.
2020-03-01 01:17:28,782 - freqtrade.freqtradebot - INFO - Buy signal found: about create a new trade with stake_amount: 108 ...
2020-03-01 01:17:28,782 - freqtrade.freqtradebot - INFO - Executing Buy for DOGEHEDGE/USD
2020-03-01 01:17:28,782 - freqtrade.freqtradebot - INFO - Using Last Ask / Last Price
2020-03-01 01:17:28,859 - freqtrade.commands.trade_commands - INFO - worker found ... calling exit
2020-03-01 01:17:28,859 - freqtrade.rpc.rpc_manager - INFO - Sending rpc message: {'type': status, 'status': 'process died'}
2020-03-01 01:17:29,229 - freqtrade.freqtradebot - INFO - Cleaning up modules ...
2020-03-01 01:17:29,229 - freqtrade.rpc.rpc_manager - INFO - Cleaning up rpc modules ...
^[[C^[[C2020-03-01 01:17:52,307 - freqtrade - ERROR - Fatal exception!
Traceback (most recent call last):
  File "/home/dante/freqtrade/freqtrade/main.py", line 36, in main
    return_code = args['func'](args)
  File "/home/dante/freqtrade/freqtrade/commands/trade_commands.py", line 20, in start_trading
    worker.run()
  File "/home/dante/freqtrade/freqtrade/worker.py", line 66, in run
    state = self._worker(old_state=state)
  File "/home/dante/freqtrade/freqtrade/worker.py", line 104, in _worker
    self._throttle(func=self._process_running, throttle_secs=self._throttle_secs)
  File "/home/dante/freqtrade/freqtrade/worker.py", line 125, in _throttle
    result = func(*args, **kwargs)
  File "/home/dante/freqtrade/freqtrade/worker.py", line 139, in _process_running
    self.freqtrade.process()
  File "/home/dante/freqtrade/freqtrade/freqtradebot.py", line 151, in process
    self.enter_positions()
  File "/home/dante/freqtrade/freqtrade/freqtradebot.py", line 217, in enter_positions
    trades_created += self.create_trade(pair)
  File "/home/dante/freqtrade/freqtrade/freqtradebot.py", line 413, in create_trade    return self.execute_buy(pair, stake_amount)
  File "/home/dante/freqtrade/freqtrade/freqtradebot.py", line 453, in execute_buy
    buy_limit_requested = self.get_buy_rate(pair, True)
  File "/home/dante/freqtrade/freqtrade/freqtradebot.py", line 249, in get_buy_rate    if ticker['ask'] < ticker['last']:
TypeError: '<' not supported between instances of 'float' and 'NoneType'
```

### PR DESCRIPTION
**Pull request by xmatthias**
## Summary
Since we've had this discussion often - we should simply make the "side" of the rate optional.
This means - it's configurable if the bot looks at the ASK or BID side for both buying and selling.

This does change the default behavior of the "side" it looks at - however the main problem with current behaviour is that as #1059 rightly pointed out - we're creating limit orders, but mostly still pay taker fees (which are often higher than maker fees).
Orderbook did work like this already (already looked at this side) - but "regular" mode did not.

closes #2907
closes #2637
closes #1667
closes #1059
closes #3005 (allow last value to be None)

## Quick changelog

- introduce `bid_strategy.price_side` and `ask_strategy.price_side`
  - valid values are `"ask"` and `"bid"` - which are automatically converted to `asks` and `bids`if orderbook is used.
- change default to look at the "correct" side of the orderbook (bids for buying, asks for selling)
- sell-order loop support through generator (so the orderbook is only queried once).

### CODE REVIEW

File: freqtrade/freqtradebot.py [hunk 1]
```diff
@@ -674,12 +684,12 @@ def handle_trade(self, trade: Trade) -> bool:
...
-
+            order_book = self._order_book_gen(trade.pair, f"{config_ask_strategy['price_side']}s",
+                                              order_book_min=order_book_min,
+                                              order_book_max=order_book_max)
             for i in range(order_book_min, order_book_max + 1):
-                order_book_rate = order_book['asks'][i - 1][0]
-                logger.debug('  order book asks top %s: %0.8f', i, order_book_rate)
-                sell_rate = order_book_rate
+                sell_rate = next(order_book)
+                logger.debug('  order book asks top %s: %0.8f', i, sell_rate)
```
Comment (hroff-1902) [inline freqtrade/freqtradebot.py]: these are not 'asks top' anymore, these are 'side's top N entries'

**Comment by hroff-1902** [comment]
I would like to make the sides non-required, with default values: 'bid' for bid_strategy and 'ask' for ask_strategy.

What do you think of this?

**Comment by xmatthias** [comment]
> I would like to make the sides non-required, with default values: 'bid' for bid_strategy and 'ask' for ask_strategy.
> What do you think of this?

Please look at `constants.py` ... guess how i set it up ... ?

File: docs/configuration.md [hunk 2]
```diff
@@ -461,23 +463,71 @@ Orderbook `bid` (buy) side depth is then divided by the orderbook `ask` (sell) s
...
+...
+```
+
+If `ask_strategy.price_side` is set to `"ask"`, then the bot will use 101 as selling price.  
+In line with that, if `ask_strategy.price_side` is set to `"bid"`, then the bot will use 99 as selling price.
+
 #### Sell price with Orderbook enabled
 
-When selling with the orderbook enabled (`ask_strategy.use_order_book=True`), Freqtrade fetches the `ask_strategy.order_book_max` entries in the orderbook. Then each of the orderbook steps between `ask_strategy.order_book_min` and `ask_strategy.order_book_max` on the `ask` orderbook side are validated for a profitable sell-possibility based on the strategy configuration and the sell order is placed at the first profitable spot.
+When selling with the orderbook enabled (`ask_strategy.use_order_book=True`), Freqtrade fetches the `ask_strategy.order_book_max` entries in the orderbook. Then each of the orderbook steps between `ask_strategy.order_book_min` and `ask_strategy.order_book_max` on the configured orderbook side are validated for a profitable sell-possibility based on the strategy configuration and the sell order is placed at the first profitable spot.
```
Comment (hroff-1902) [inline docs/configuration.md]:
```suggestion
When selling with the orderbook enabled (`ask_strategy.use_order_book=True`), Freqtrade fetches the `ask_strategy.order_book_max` entries in the orderbook. Then each of the orderbook steps between `ask_strategy.order_book_min` and `ask_strategy.order_book_max` on the configured orderbook side are validated for a profitable sell-possibility based on the strategy configuration (`minimal_roi` conditions) and the sell order is placed at the first profitable spot.
```

File: docs/configuration.md [hunk 3]
```diff
@@ -461,23 +463,71 @@ Orderbook `bid` (buy) side depth is then divided by the orderbook `ask` (sell) s
...
+99   # bid
+98
+97
+...
+```
+
+If `bid_strategy.price_side` is set to `"bid"`, then the bot will use 99 as buying price.  
+In line with that, if `bid_strategy.price_side` is set to `"ask"`, then the bot will use 101 as buying price.
+
+Using `ask` price often guarantees quicker filled orders, but the bot can also end up paying more than what would have been necessary.
```
[the remark this answers is not in the document]
Comment (xmatthias) [reply docs/configuration.md]: i did slightly reword this to make it clearer - also i think splitting this out into 3 sentences makes it easier to understandable (i hope) ... overly long sentences are sometimes hard to digets for non-native english speakers.
